File: scala/src/main/scala/bron_kerbosch1.scala
import scala.collection.{immutable, mutable}

class bron_kerbosch1[Vertex] extends bron_kerbosch_algorithm[Vertex] {
  def explore(
      graph: UndirectedGraph[Vertex],
      reporter: immutable.Iterable[Vertex] => Unit
  ): Unit = {
    val candidates = graph.connected_vertices().to(mutable.Set)
    if (candidates.nonEmpty)
      visit(graph, reporter, candidates, mutable.Set(), immutable.List())
  }

  def visit(
      graph: UndirectedGraph[Vertex],
      reporter: immutable.Iterable[Vertex] => Unit,
      candidates: mutable.Set[Vertex],
      excluded: mutable.Set[Vertex],
      clique_in_progress: immutable.List[Vertex]
  ): Unit = {
    assert(candidates.forall(v => graph.degree(v) > 0))
    assert(excluded.forall(v => graph.degree(v) > 0))
    assert(util.are_disjoint(candidates, excluded))
    assert(candidates.nonEmpty)
    while (candidates.nonEmpty) {
      val v = candidates.head
      candidates -= v
      val neighbours = graph.neighbours(v)
      val neighbouring_candidates = util.intersect(neighbours, candidates)
      if (neighbouring_candidates.nonEmpty) {
        val neighbouring_excluded = util.intersect(neighbours, excluded)
        visit(
          graph,
          reporter,
          neighbouring_candidates.to(mutable.Set),
          neighbouring_excluded.to(mutable.Set),
          v :: clique_in_progress
        )
      } else if (util.are_disjoint(neighbours, excluded)) {
        reporter(v :: clique_in_progress)
      }
      excluded += v
    }
  }
}

File: kotlin/src/main/be.steinsomers.bron_kerbosch/BronKerbosch1.kt
// Naive Bron-Kerbosch algorithm
package be.steinsomers.bron_kerbosch

class BronKerbosch1 : BronKerboschAlgorithm {
    override fun explore(graph: UndirectedGraph, cliqueConsumer: CliqueConsumer) {
        val candidates: MutableSet<Vertex> = graph.connectedVertices().toCollection(HashSet())
        val excluded: MutableSet<Vertex> = HashSet(candidates.size)
        visit(
            graph = graph, cliqueConsumer = cliqueConsumer,
            candidates = candidates, excluded = excluded,
            clique = Clique.empty()
        )
    }

    companion object {
        private fun visit(
            graph: UndirectedGraph, cliqueConsumer: CliqueConsumer,
            candidates: MutableSet<Vertex>, excluded: MutableSet<Vertex>,
            clique: Clique
        ) {
            Debug.assert { candidates.all(graph::hasDegree) }
            Debug.assert { excluded.all(graph::hasDegree) }
            Debug.assert { Util.areDisjoint(candidates, excluded) }
            while (candidates.isNotEmpty()) {
                val v = Util.popArbitrary(candidates)
                val neighbours = graph.neighbours(v)
                val neighbouringCandidates = Util.intersect(candidates, neighbours)
                if (neighbouringCandidates.isNotEmpty()) {
                    val neighbouringExcluded = Util.intersect(excluded, neighbours)
                    visit(
                        graph = graph, cliqueConsumer = cliqueConsumer,
                        candidates = neighbouringCandidates, excluded = neighbouringExcluded,
                        clique = clique.plus(v)
                    )
                } else if (clique.size() + 1 >= cliqueConsumer.minSize && Util.areDisjoint(
                        excluded,
                        neighbours
                    )
                ) {
                    cliqueConsumer.accept(clique.plus(v))
                }
                excluded.add(v)
            }
        }
    }
}
